### PR TITLE
Fix infinite loop and use faster size computation

### PR DESCRIPTION
**Pull request by amberarrow**
@vrozov Please take a look

Fixed infinite loop for large size arguments, added checks for unreasonable arguments, used faster method to compute buffer size, added unit tests.

### CODE REVIEW

File: src/main/java/com/datatorrent/netlet/util/CircularBuffer.java [hunk 1]
```diff
@@ -52,13 +56,14 @@
   @SuppressWarnings("unchecked")
   public CircularBuffer(int n, int spin)
   {
-    int i = 1;
-    while (i < n) {
-      i <<= 1;
-    }
+    if (n < 0 || n > MAX_SIZE) throw new RuntimeException(String.format("Error: value too large: %d%n", n));
```
Comment (vrozov) [inline src/main/java/com/datatorrent/netlet/util/CircularBuffer.java]: IllegalArgumentException?

File: pom.xml [hunk 2]
```diff
@@ -220,6 +220,14 @@
           <logViolationsToConsole>${checkstyle.console}</logViolationsToConsole>
         </configuration>
       </plugin>
+      <plugin>
```
Comment (vrozov) [inline pom.xml]: Do not rely on the availability of 8gb of memory.

File: src/main/java/com/datatorrent/netlet/util/CircularBuffer.java (hunk 1 above)
Comment (vrozov) [inline src/main/java/com/datatorrent/netlet/util/CircularBuffer.java]: The error message assumes that value is positive, what if n < 0?

Comment (vrozov) [inline src/main/java/com/datatorrent/netlet/util/CircularBuffer.java]: %n?

File: src/main/java/com/datatorrent/netlet/util/CircularBuffer.java [hunk 3]
```diff
@@ -52,13 +56,14 @@
...
+
+    final int size = (0 == n) ? 1
+            : (0 == (n & (n - 1))) ? n
+            : (1 << (32 - Integer.numberOfLeadingZeros(n)));
 
-    buffer = (T[])new Object[i];
-    buffermask = i - 1;
+    buffer = (T[])new Object[size];
```
Comment (vrozov) [inline src/main/java/com/datatorrent/netlet/util/CircularBuffer.java]: `this((T[])new Object[size], size - 1, spin);`?

File: src/test/java/com/datatorrent/netlet/util/CircularBufferTest.java [hunk 4]
```diff
@@ -52,6 +52,57 @@ public void tearDown()
...
+   */
+  @Test
+  public void testLargeSize()
+  {
+    try {
+      final int size = 1 + CircularBuffer.MAX_SIZE;
+      CircularBuffer<Integer> instance = new CircularBuffer<Integer>(size);
+      assert false;
+    } catch (RuntimeException e) {    // expected
+      System.out.format("Got expected exception: %s%n", e);
```
Comment (vrozov) [inline src/test/java/com/datatorrent/netlet/util/CircularBufferTest.java]: No System.out, use log4j if necessary

File: src/test/java/com/datatorrent/netlet/util/CircularBufferTest.java [hunk 5]
```diff
@@ -52,6 +52,57 @@ public void tearDown()
...
   /**
+   * Test larger than allowed size of buffer; we expect an exception
+   */
+  @Test
+  public void testLargeSize()
+  {
+    try {
+      final int size = 1 + CircularBuffer.MAX_SIZE;
+      CircularBuffer<Integer> instance = new CircularBuffer<Integer>(size);
+      assert false;
```
Comment (vrozov) [inline src/test/java/com/datatorrent/netlet/util/CircularBufferTest.java]: No assert, use junit.

File: src/test/java/com/datatorrent/netlet/util/CircularBufferTest.java [hunk 6]
```diff
@@ -52,6 +52,57 @@ public void tearDown()
...
 
   /**
+   * Test larger than allowed size of buffer; we expect an exception
+   */
+  @Test
+  public void testLargeSize()
+  {
+    try {
+      final int size = 1 + CircularBuffer.MAX_SIZE;
+      CircularBuffer<Integer> instance = new CircularBuffer<Integer>(size);
```
Comment (vrozov) [inline src/test/java/com/datatorrent/netlet/util/CircularBufferTest.java]: `new CircularBuffer<Integer>(1 + CircularBuffer.MAX_SIZE);`? Both ways OK.

File: src/test/java/com/datatorrent/netlet/util/CircularBufferTest.java [hunk 7]
```diff
@@ -52,6 +52,57 @@ public void tearDown()
...
+  private void testSize(final int size)
+  {
+    CircularBuffer<Integer> instance = new CircularBuffer<Integer>(size);
+
+    // get size the old way and check that it matches
+    int i = 1;
+    while (i < size) {
+      i <<= 1;
+    }
+    assert i == instance.capacity();
```
Comment (vrozov) [inline src/test/java/com/datatorrent/netlet/util/CircularBufferTest.java]: no assert, use junit

File: src/test/java/com/datatorrent/netlet/util/CircularBufferTest.java [hunk 8]
```diff
@@ -52,6 +52,57 @@ public void tearDown()
...
+
+    // check that the instance is actually functional
+    instance.add(99);
+    final int v = instance.remove();
+    assert 99 == v;
+
+    instance = null;
```
Comment (vrozov) [inline src/test/java/com/datatorrent/netlet/util/CircularBufferTest.java]: why is it necessary to set instance to null?

File: src/test/java/com/datatorrent/netlet/util/CircularBufferTest.java [hunk 9]
```diff
@@ -52,6 +52,57 @@ public void tearDown()
   }
 
   /**
+   * Test larger than allowed size of buffer; we expect an exception
+   */
+  @Test
+  public void testLargeSize()
+  {
+    try {
+      final int size = 1 + CircularBuffer.MAX_SIZE;
```
Comment (vrozov) [inline src/test/java/com/datatorrent/netlet/util/CircularBufferTest.java]: static import MAX_SIZE?

File: src/test/java/com/datatorrent/netlet/util/CircularBufferTest.java [hunk 10]
```diff
@@ -52,6 +52,57 @@ public void tearDown()
...
+    // get size the old way and check that it matches
+    int i = 1;
+    while (i < size) {
+      i <<= 1;
+    }
+    assert i == instance.capacity();
+
+    // check that the instance is actually functional
+    instance.add(99);
+    final int v = instance.remove();
```
Comment (vrozov) [inline src/test/java/com/datatorrent/netlet/util/CircularBufferTest.java]: use poll() and peek(). Check that the instance is empty after the last expected element is deleted from the CircularBuffer.

File: src/test/java/com/datatorrent/netlet/util/CircularBufferTest.java [hunk 11]
```diff
@@ -52,6 +52,57 @@ public void tearDown()
...
+  }
+
+ /**
+  * Test values of the form 2**k and (1 + 2**k) as the size of the buffer
+  */
+  @Test
+  public void testSize()
+  {
+    final int max_size = CircularBuffer.MAX_SIZE;
+    for (int size = 0; size < max_size; size <<= 1) {
```
Comment (vrozov) [inline src/test/java/com/datatorrent/netlet/util/CircularBufferTest.java]: use MAX_SIZE

File: src/test/java/com/datatorrent/netlet/util/CircularBufferTest.java [hunk 12]
```diff
@@ -52,6 +52,57 @@ public void tearDown()
...
+      testSize(size);
+      testSize(1 + size);
```
Comment (vrozov) [inline src/test/java/com/datatorrent/netlet/util/CircularBufferTest.java]: check size + 1 and size - 1.

File: src/main/java/com/datatorrent/netlet/util/CircularBuffer.java (hunk 1 above)
Comment (amberarrow) [inline src/main/java/com/datatorrent/netlet/util/CircularBuffer.java]: %n is the (portable) Java-ism for a newline; I use it to separate the message from the rest of the text.

Comment (vrozov) [inline src/main/java/com/datatorrent/netlet/util/CircularBuffer.java]: I don't think that a new line is necessary or desired in the exception message.

File: src/main/java/com/datatorrent/netlet/util/CircularBuffer.java (hunk 3 above)
Comment (amberarrow) [inline src/main/java/com/datatorrent/netlet/util/CircularBuffer.java]: Call to another constructor must be first statement of function body, so cannot do this.

**Comment by amberarrow** [comment]
@vrozov Travis is happy; please merge

**Pull request merged**